Improve logging on ONNX symbolic helpers

thiagocrepaldi opened the pull request:
Signed-off-by: Thiago Crepaldi <thiago.crepaldi@microsoft.com>

Fixes #ISSUE_NUMBER

Commit:
Improve logging on ONNX symbolic helpers

Signed-off-by: Thiago Crepaldi <thiago.crepaldi@microsoft.com>

BowenBao commented on test/onnx/test_operators.py:
```diff
@@ -122,6 +122,32 @@ def assertONNXRaisesRegex(self, err, reg, f, args, params=None, **kwargs):
+        _onnx_opset_version = 9
+
+        import torch.onnx.symbolic_helper as sym_help
+        # import torch.onnx.symbolic_registry as sym_registry
```
remove commented code

BowenBao commented on test/onnx/test_operators.py:
```diff
@@ -122,6 +122,32 @@ def assertONNXRaisesRegex(self, err, reg, f, args, params=None, **kwargs):
+
+        @parse_args("v")
+        def cat(g, tensor_list, dim):
+            print('#' * 1024)
```
remove print

BowenBao commented on torch/onnx/symbolic_helper.py:
```diff
@@ -173,7 +173,12 @@ def decorator(fn):
         def wrapper(g, *args, **kwargs):
             # some args may be optional, so the length may be smaller
-            assert len(arg_descriptors) >= len(args)
+            # if fn.__name__ == 'cat':
```
remove commented code

BowenBao commented on torch/onnx/symbolic_helper.py:
```diff
@@ -173,7 +173,12 @@ def decorator(fn):
+            #     import pdb; pdb.set_trace()
+            assert len(arg_descriptors) >= len(args),\
+                f"A mismatch between the number of arguments ({len(args)}) and"\
+                f" their descriptors ({len(arg_descriptors)}) was found at function '{fn.__name__}'"
```
curious would it be possible if fn does not have attribute `__name__`?

BowenBao replied:
On top of that, I feel we could provide actionable suggestions in this message. e.g. submit bug for official symbolic implementation, or fix custom symbolic implementation. Perhaps even differentiate if the symbolic is custom or not, but that could be a bit too complicated.

thiagocrepaldi replied:
There will always be a `__name__`. In fact, the main goal of `@wraps(fn)` is to copy the `fn.__name__` and `fn.__docstring__` into `wrapper`'s

thiagocrepaldi replied:
> On top of that, I feel we could provide actionable suggestions in this message. e.g. submit bug for official symbolic implementation, or fix custom symbolic implementation. Perhaps even differentiate if the symbolic is custom or not, but that could be a bit too complicated.

One of the issues in being actionable is that the issue might be happening on custom ops and contrib ops. This is exactly what triggered this PR. ONNX Runtime custom symbolics differed from PyTorch's after migrating from 1.9.0 to 1.10

I will try rewrite the error so it explicitly says the symbolic is broken.

BowenBao reviewed (changes requested):
Please check lint failures.

Commit:
Fix CI and address comments

Signed-off-by: Thiago Crepaldi <thiago.crepaldi@microsoft.com>

BowenBao commented on torch/onnx/symbolic_helper.py:
```diff
@@ -173,7 +173,14 @@ def decorator(fn):
+                "https://github.com/pytorch/pytorch/issues/new?template=bug-report.md to report this bug."
+            assert len(arg_descriptors) >= len(args),\
+                f"A mismatch between the number of arguments ({len(args)}) and"\
+                f" their descriptors ({len(arg_descriptors)}) was found at symbolic function '{fn.__name__}'. "\
```
nit: would it be possible to print the file where `fn` is at? This could be helpful figuring out which particular opset had this issue.

thiagocrepaldi replied:
I agree and I have tried before, I couldn't figure out a way that doesn't envolve the backtarce, which is printed anyway as the assert is triggered. I have tried using `inspect` and `traceback`

BowenBao commented on test/onnx/test_operators.py:
```diff
@@ -122,6 +122,33 @@ def assertONNXRaisesRegex(self, err, reg, f, args, params=None, **kwargs):
+            AssertionError,
+            lambda: torch.onnx.export(model, (x,), f, opset_version=_onnx_opset_version),
+            ("A mismatch between the number of arguments (2) and their descriptors (1) was found at symbolic function "
+             "'cat'. If you believe this is not due to custom symbolic implementation within your code oran external "
```
```suggestion
             "'cat'. If you believe this is not due to custom symbolic implementation within your code or an external "
```

BowenBao commented on test/onnx/test_operators.py:
```diff
@@ -122,6 +122,33 @@ def assertONNXRaisesRegex(self, err, reg, f, args, params=None, **kwargs):
+            lambda: torch.onnx.export(model, (x,), f, opset_version=_onnx_opset_version),
+            ("A mismatch between the number of arguments (2) and their descriptors (1) was found at symbolic function "
+             "'cat'. If you believe this is not due to custom symbolic implementation within your code oran external "
+             "library, please file an issue at https://github.com/pytorch/pytorch/issues/new?template=bug-report.md to "
```
```suggestion
             "library, please file an issue at https://github.com/pytorch/pytorch/issues/new?template=bug-report.yml to "
```

thiagocrepaldi replied:
Interesting, I copied and pasted this from another file. There is a typo there too

BowenBao commented on test/onnx/test_operators.py:
```diff
@@ -122,6 +122,33 @@ def assertONNXRaisesRegex(self, err, reg, f, args, params=None, **kwargs):
         with self.assertRaisesRegex(err, reg):
             export_to_pbtxt(m, args, **kwargs)
 
+    def test_bad_symbolic_registration(self):
```
nit: `test_utility_funs.py` would be a more suitable place for this kind of test.

BowenBao replied:
The similar test is 4 years old .. Back at the time there wasn't other test suites for different purposes.

BowenBao reviewed (approved):
LGTM, some minor nits and comments.

Commit:
Update test/onnx/test_operators.py

Co-authored-by: Bowen Bao <imbowenbao@outlook.com>

Commit:
Address comments

Signed-off-by: Thiago Crepaldi <thiago.crepaldi@microsoft.com>

BowenBao commented on torch/nn/modules/module.py:
```diff
@@ -556,7 +556,7 @@ def get_extra_state(self) -> Any:
         raise RuntimeError(
             "Reached a code path in Module.get_extra_state() that should never be called. "
-            "Please file an issue at https://github.com/pytorch/pytorch/issues/new?template=bug-report.md "
+            "Please file an issue at https://github.com/pytorch/pytorch/issues/new?template=bug-report.yml "
```
unfortunately by helping updating this part, this PR won't be mergeable by bot ... Leaving for you to decide

thiagocrepaldi replied:
I will create a separate PR then

thiagocrepaldi commented:
@pytorchbot merge this

rohan-varma commented:
ONNX PR CI is failing here: https://github.com/pytorch/pytorch/runs/5327758015?check_suite_focus=true and causing issues on master, planning to revert.

malfet commented:
@pytorchbot revert this

Commit:
Fix missing imports

thiagocrepaldi commented:
> ONNX PR CI is failing here: https://github.com/pytorch/pytorch/runs/5327758015?check_suite_focus=true and causing issues on master, planning to revert.

Sorry folks, during the last review update, I've moved a unit test from one file to another and forgot to do the proper imports there. Somehow CI doesn't seem to have caught it? I have fixed it and will wait for the CI to confirm everything is fine

thiagocrepaldi commented:
@pytorchbot merge this
